diffrax/_solver/srk.py

```diff
@@ -0,0 +1,740 @@
+    AbstractSpaceTimeLevyArea,
+    AbstractSpaceTimeTimeLevyArea,
+    BoolScalarLike,
+    DenseInfo,
+    FloatScalarLike,
+    IntScalarLike,
+    RealScalarLike,
+    VF,
+    Y,
+)
+from .._local_interpolation import LocalLinearInterpolation
+from .._solution import RESULTS
+from .._term import AbstractTerm, MultiTerm, ODETerm
+from .base import AbstractSolver
+
+
+if TYPE_CHECKING:
+    from typing import ClassVar as AbstractClassVar
+else:
+    from equinox import AbstractClassVar
+
+_ErrorEstimate: TypeAlias = Optional[Y]
+_SolverState: TypeAlias = None
+_CarryType: TypeAlias = tuple[PyTree[Array], PyTree[Array], PyTree[Array]]
+
+
+class AbstractStochasticCoeffs(eqx.Module):
+    a: eqx.AbstractVar[Union[Float[np.ndarray, " s"], tuple[np.ndarray, ...]]]
+    b: eqx.AbstractVar[Union[Float[np.ndarray, " s"], FloatScalarLike]]
+
+    def check(self):
+        raise NotImplementedError
+
+
+class AdditiveCoeffs(AbstractStochasticCoeffs):
+    """Coefficients for either the Brownian increment or its Lévy areas in an
+    SRK for solving additive noise SDEs.
+    """
+
+    # assuming SDE has additive noise we only need a 1-dimensional array
+    # of length s for the coefficients in front of the Brownian increment
+    # and/or Lévy areas (where s is the number of stages of the solver).
+    # This is the equivalent of the matrix a for the Brownian motion and
+    # its Lévy areas.
+    a: Float[np.ndarray, " s"]
+    b: FloatScalarLike
+
+    def check(self):
+        assert self.a.ndim == 1
+        return self.a.shape[0]
+
+
+class _AbstractGeneralCoeffs(AbstractStochasticCoeffs):
+    a: eqx.AbstractVar[tuple[np.ndarray, ...]]
+    b: eqx.AbstractVar[Float[np.ndarray, " s"]]
+
+
+class GeneralCoeffs(_AbstractGeneralCoeffs):
+    """General coefficients for either the Brownian increment or its Lévy areas in an
+    SRK for solving SDEs with any type of noise (i.e. non-additive).
+    """
+
+    a: tuple[np.ndarray, ...]
+    b: Float[np.ndarray, " s"]
+
+    def check(self):
+        assert self.b.ndim == 1
+        assert all((i + 1,) == a_i.shape for i, a_i in enumerate(self.a))
+        assert self.b.shape[0] == len(self.a) + 1
+        return self.b.shape[0]
+
+
+class GeneralCoeffsWithError(_AbstractGeneralCoeffs):
+    """General coefficients for either the Brownian increment or its Lévy areas in an
+    SRK for solving SDEs with any type of noise (i.e. non-additive). Use this subclass
+    when the SRK provides an embedded method for error estimation.
+    """
+
+    a: tuple[np.ndarray, ...]
+    b: Float[np.ndarray, " s"]
+    b_error: Float[np.ndarray, " s"]
+
+    def check(self):
+        assert self.b.ndim == 1
+        assert all((i + 1,) == a_i.shape for i, a_i in enumerate(self.a))
+        assert self.b.shape[0] == len(self.a) + 1
+        assert self.b_error.ndim == 1
+        assert self.b_error.shape == self.b.shape
+        return self.b.shape[0]
+
+
+COEFFS = TypeVar("COEFFS", bound=AbstractStochasticCoeffs)
```

Comment:
Nit: `_Coeffs` would be the usual way of writing a private (underscore) typevar (CamelCase)